From 6f42216923b6d1bb4924d6e2c9117a226d40c2f3 Mon Sep 17 00:00:00 2001
From: maximilian attems <max@stro.at>
Date: Sun, 7 Sep 2008 10:01:28 +0200
Subject: [PATCH 01/21] [klibc] add strtok_r()

reentrant version of strtok() was missing in klibc,
use the current strtok() implementation and make
use of it in strtok() itself.

Signed-off-by: maximilian attems <max@stro.at>
---
 usr/include/string.h |    1 +
 usr/klibc/Kbuild     |    1 +
 usr/klibc/strtok.c   |    9 +--------
 usr/klibc/strtok_r.c |   13 +++++++++++++
 4 files changed, 16 insertions(+), 8 deletions(-)
 create mode 100644 usr/klibc/strtok_r.c

diff --git a/usr/include/string.h b/usr/include/string.h
index ae8270e..0c8c046 100644
--- a/usr/include/string.h
+++ b/usr/include/string.h
@@ -44,5 +44,6 @@ __extern char *strsep(char **, const char *);
 __extern size_t strspn(const char *, const char *);
 __extern char *strstr(const char *, const char *);
 __extern char *strtok(char *, const char *);
+__extern char *strtok_r(char *, const char *, char **);
 
 #endif				/* _STRING_H */
diff --git a/usr/klibc/Kbuild b/usr/klibc/Kbuild
index c8eabf9..eb04267 100644
--- a/usr/klibc/Kbuild
+++ b/usr/klibc/Kbuild
@@ -36,6 +36,7 @@ klib-y := vsnprintf.o snprintf.o vsprintf.o sprintf.o \
 	  strncat.o strlcpy.o strlcat.o \
 	  strstr.o strncmp.o strncpy.o strrchr.o \
 	  strxspn.o strspn.o strcspn.o strpbrk.o strsep.o strtok.o \
+	  strtok_r.o \
 	  fnmatch.o \
 	  gethostname.o getdomainname.o getcwd.o \
 	  seteuid.o setegid.o \
diff --git a/usr/klibc/strtok.c b/usr/klibc/strtok.c
index c2671af..6b169a1 100644
--- a/usr/klibc/strtok.c
+++ b/usr/klibc/strtok.c
@@ -8,12 +8,5 @@ char *strtok(char *s, const char *delim)
 {
 	static char *holder;
 
-	if (s)
-		holder = s;
-
-	do {
-		s = strsep(&holder, delim);
-	} while (s && !*s);
-
-	return s;
+	return strtok_r(s, delim, &holder);
 }
diff --git a/usr/klibc/strtok_r.c b/usr/klibc/strtok_r.c
new file mode 100644
index 0000000..695d516
--- /dev/null
+++ b/usr/klibc/strtok_r.c
@@ -0,0 +1,13 @@
+#include <string.h>
+
+char *strtok_r(char *s, const char *delim, char **holder)
+{
+	if (s)
+		*holder = s;
+
+	do {
+		s = strsep(holder, delim);
+	} while (s && !*s);
+
+	return s;
+}
-- 
1.6.5

